ghcide/test/exe/Main.hs

```diff
@@ -1464,6 +1476,106 @@ suggestImportTests = testGroup "suggest import actions"
+                "Use AVec for Vec, hiding other imports"
+                "HideType.hs.expected.A"
+        , testCase "EVec" $
+            compareTwo
+                "HideType.hs"
+                "Use EVec for Vec, hiding other imports"
+                "HideType.hs.expected.E"
+        ]
+    ]
+  , testGroup "Qualify strategy"
+    [ testCase "won't suggest full name for qualified module" $
+      withHideFunction $ \_ actions -> do
+        liftIO $
+            assertBool "EVec.fromList must not be suggested" $
+                "Replace with qualified: EVec.fromList" `notElem`
+                [ actionTitle
+                | CACodeAction CodeAction { _title = actionTitle } <- actions
+                ]
+        liftIO $
+            assertBool "EVec.++ must not be suggested" $
+                "Replace with qualified: EVec.++" `notElem`
+                [ actionTitle
+                | CACodeAction CodeAction { _title = actionTitle } <- actions
+                ]
+    , testGroup "fromList"
+        [ testCase "EVec" $
+            compareHideFunctionTo
+                "Replace with qualified: E.fromList"
+                "HideFunction.hs.expected.qualified.fromList.E"
+        ]
+    , testGroup "(++)"
+        [ testCase "Prelude" $
+            compareHideFunctionTo
+                "Replace with qualified: Prelude.++"
+                "HideFunction.hs.expected.qualified.append.Prelude"
+        ]
+    ]
+  ]
+  where
+    hidingDir = "test/data/hiding"
+    compareTwo original cmd expected =
+        withTarget original $ \doc actions -> do
+            expected <- liftIO $
+                readFileUtf8 (hidingDir </> expected)
+            action <- liftIO $ pickActionWithTitle cmd actions
+            executeCodeAction action
+            contentAfterAction <- documentContents doc
+            liftIO $ T.replace "\r\n" "\n" expected @=? contentAfterAction
+    compareHideFunctionTo = compareTwo "HideFunction.hs"
+    withTarget file k = runInDir hidingDir $ do
+        doc <- openDoc file "haskell"
+        void (skipManyTill anyMessage message
+            :: Session WorkDoneProgressEndNotification)
+        void waitForDiagnostics
+        contents <- documentContents doc
+        let range = Range (Position 0 0) (Position (length $ T.lines contents) 0)
+        actions <- waitForAtLeatOneAction 0.5 4 doc range
+        k doc actions
+    withHideFunction = withTarget ("HideFunction" <.> "hs")
+
+waitForAtLeatOneAction ::
+    -- | Waiting interval
+    Double ->
+    -- | Maximum # of retry (0 for no retry at ll)
+    Natural ->
+    TextDocumentIdentifier ->
+    Range ->
+    Session [CAResult]
+waitForAtLeatOneAction wait count doc range = go count []
+    where
+        go !remain !acc = do
+            liftIO $ sleep wait
+            actions <- getCodeActions doc range
+            if not (null actions) || remain <= 0
+                then pure $ acc ++ actions
+                else go (remain - 1) (acc ++ actions)
```

Comment:
Are you sure that we need this? 

I believe that ghcide currently returns all the code actions at the same time

The only case where a recursive call would ever be needed is if a test is expecting too many code actions. This will cause an unnecessary delay and go undetected, slowing down the test suite forever

Reply:
This was not at first introduced. But it seems some test-cases introduced in this PR, especially `Use AVec for Vec, hiding other imports` and `Use EVec for Vec, ...`, continually fail in CI complaining that `Found no matching actions for "Use AVec for Vec, hiding other imports" in []` (see, for example [this log](https://github.com/haskell/haskell-language-server/runs/1797000004?check_suite_focus=true)).
Once I tried to increase `sleep` duration from `0.5` to `1.0`, `1.5` and then `2.0`. But it still fails by accident.
I can, of course, increase sleep duration further, but it also introduces additional overhead to other tests related to disambiguation actions.

Hence, I decided to call `getCodeActions` repeatedly until any action becomes available.
It can be the case that some bug is accidentally introduced and no code action is generated, so I limit the number of retries. This is the intention of this hack.

Reply:
The problem is that `waitForDiagnostics` doesn't wait for *all* the diagnostics.
You need to wait for diagnostics on the source file where you expect the code action.
Use `expectDiagnostics` instead

Reply:
Thank you for your suggestion! I've just switched to `expectDiagnostics` and changed code to open documents in separate isolated directory. Everything works fine 😄